parser.py

```diff
@@ -1,17 +1,49 @@
 #!/bin/env python
 import argparse
 import pprint
+import re
+from abc import abstractmethod
+from typing import override
+
 import grammar
 import visitor
 
 
+class Preprocessor:
+    @abstractmethod
+    def process(self, program: str):
+        pass
+
+
+class CommentsPreprocessor(Preprocessor):
+
+    @override
+    def process(self, program: str):
+        cleaned_lines = filter(
+            lambda cleaned_line: cleaned_line,
+            map(lambda line: re.sub(r'#.*$', '', line), program.splitlines()))
+        return "\n".join(cleaned_lines)
+
+
+class PreprocessorPipeline:
+    def __init__(self, preprocessor_actions ):
+        self.preprocessor_actions = preprocessor_actions
+
+    def process(self, program: str):
+        for action in self.preprocessor_actions:
+            program = action.process(program)
+        return program
+
+
 class ImperivmParser:
     def __init__(self, grammar=grammar.imperivm, visitor=visitor.ImperivmVisitor()):
         self.grammar = grammar
         self.visitor = visitor
+        self.preprocessor = PreprocessorPipeline([CommentsPreprocessor()])
```

Comment:
To allow dependency injection this should be received in a parameter, same as visitor and grammar. You can provide a default value for ease of use.